hubspot-style/src/main/java/com/hubspot/immutables/utils/WireSafeEnum.java

```diff
@@ -0,0 +1,151 @@
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.databind.BeanProperty;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.deser.ContextualDeserializer;
+import com.hubspot.immutables.utils.WireSafeEnum.Deserializer;
+
+@JsonDeserialize(using = Deserializer.class)
+public final class WireSafeEnum<T extends Enum<T>> {
+  private static final Map<Class<?>, Map<?, WireSafeEnum<?>>> ENUM_LOOKUP_CACHE =
+      new ConcurrentHashMap<>();
+  private static final Map<Class<?>, Map<String, WireSafeEnum<?>>> STRING_LOOKUP_CACHE =
+      new ConcurrentHashMap<>();
+
+  private final Class<T> enumType;
+  private final String stringValue;
+  private final Optional<T> enumValue;
+
+  private WireSafeEnum(Class<T> enumType, T value) {
+    this.enumType = enumType;
+    this.stringValue = value.name();
+    this.enumValue = Optional.of(value);
+  }
+
+  private WireSafeEnum(Class<T> enumType, String stringValue) {
+    this.enumType = enumType;
+    this.stringValue = stringValue;
+    this.enumValue = Optional.empty();
+  }
+
+  @SuppressWarnings("unchecked")
+  public static <T extends Enum<T>> WireSafeEnum<T> of(T value) {
+    Class<T> enumType = (Class<T>) value.getClass();
+    ensureEnumCacheInitialized(enumType);
+    return (WireSafeEnum<T>) ENUM_LOOKUP_CACHE.get(enumType).get(value);
+  }
+
+  @SuppressWarnings("unchecked")
+  public static <T extends Enum<T>> WireSafeEnum<T> of(Class<T> enumType, String value) {
+    ensureStringCacheInitialized(enumType);
+    WireSafeEnum<?> cached = STRING_LOOKUP_CACHE.get(enumType).get(value);
+    if (cached == null) {
+      return new WireSafeEnum<>(enumType, value);
+    } else {
+      return (WireSafeEnum<T>) cached;
+    }
+  }
+
+  public Class<T> enumType() {
+    return enumType;
+  }
+
+  @JsonValue
+  public String asString() {
+    return stringValue;
+  }
+
+  public Optional<T> asEnum() {
+    return enumValue;
+  }
+
+  private static <T extends Enum<T>> void ensureEnumCacheInitialized(Class<T> enumType) {
+    if (!ENUM_LOOKUP_CACHE.containsKey(enumType)) {
+      initializeCache(enumType);
+    }
+  }
+
+  private static <T extends Enum<T>> void ensureStringCacheInitialized(Class<T> enumType) {
+    if (!STRING_LOOKUP_CACHE.containsKey(enumType)) {
+      initializeCache(enumType);
+    }
+  }
+
+  private static <T extends Enum<T>> void initializeCache(Class<T> enumType) {
+    Map<T, WireSafeEnum<?>> enumMap = new EnumMap<>(enumType);
+    Map<String, WireSafeEnum<?>> stringMap = new HashMap<>();
```

Comment:
Can we size this?

Reply:
Updated to size it, but sizing of hash maps is always a pain. Wish we could use Guava's ImmutableMap 😢